Tests/PenguinStructuresTests/TupleTests.swift

```diff
@@ -0,0 +1,167 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import XCTest
+import PenguinStructures
+
+// ************************************************************************
+// A demonstration that we can write straightforward algorithms over tuples
+// ************************************************************************
+
+/// Function-like types that can be passed any argument type.
+protocol GenericFunction {
+  associatedtype Result
+  func callAsFunction<T>(_: T) -> Result
+}
+
+/// Returns `[T].self`, given an argument of type `T`
+struct ArrayType: GenericFunction {
+  func callAsFunction<T>(_ x: T) -> Any.Type {
+    [T].self
+  }
+}
+
+struct X<T>: DefaultInitializable, Equatable {}
+
+extension TupleProtocol {
+  /// Returns `start` reduced via `reducer` into the result of calling `mapper` 
+  /// on each element.
+  func mapReduce<M: GenericFunction, R>(
+    _ start: R, _ mapper: M, _ reducer: (R, M.Result)->R
+  ) -> R {
+    if Self.self == Empty.self { return start }
+    let next = reducer(start, mapper(head))
+    return tail.mapReduce(next, mapper, reducer)
+  }
+}
+
+class TupleTests: XCTestCase {
+  func test_mapReduce() {
+    XCTAssertEqual(
+      Tuple(2, 3.4, "Bar").mapReduce([], ArrayType()) { $0 + [$1] }
+        .lazy.map(ObjectIdentifier.init),
+      [[Int].self, [Double].self, [String].self]
+        .lazy.map(ObjectIdentifier.init)
+    )
+  }
+
+  func test_head() {
+    XCTAssertEqual(Tuple(0).head, 0)
+    XCTAssertEqual(Tuple(0.0, 1).head, 0.0)
+    XCTAssertEqual(Tuple("foo", 0.0, 1).head, "foo")
+  }
+  
+  func test_tail() {
+    XCTAssertEqual(Tuple(0).tail, Empty())
+    XCTAssertEqual(Tuple(0.0, 1).tail, Tuple(1))
+    XCTAssertEqual(Tuple("foo", 0.0, 1).tail, Tuple(0.0, 1))
+  }
+
+  func test_count() {
+    typealias I = Int
+    XCTAssertEqual(Tuple0.count, 0)
+    XCTAssertEqual(Tuple1<I>.count, 1)
+    XCTAssertEqual(Tuple2<I, I>.count, 2)
+    XCTAssertEqual(Tuple3<I, I, I>.count, 3)
+  }
+
+  func test_DefaultInitializable() {
+    XCTAssertEqual(Tuple0(), Empty())
+    XCTAssertEqual(Tuple1<X<Int>>(), Tuple(.init()))
+    XCTAssertEqual(Tuple2<X<Int>, X<String>>(), Tuple(.init(), .init()))
+  }
+  
+  func test_Equatable() {
+    XCTAssertEqual(Tuple0(), Tuple0())
+    
+    XCTAssertEqual(Tuple(1), Tuple(1))
+    XCTAssertNotEqual(Tuple(1), Tuple(2))
+    
+    XCTAssertEqual(Tuple(1, 2), Tuple(1, 2))
+    XCTAssertNotEqual(Tuple(0, 2), Tuple(1, 2))
+    XCTAssertNotEqual(Tuple(1, "2"), Tuple(1, "XXX"))
+    
+    XCTAssertEqual(Tuple(1, 2.3, "4"), Tuple(1, 2.3, "4"))
+    XCTAssertNotEqual(Tuple(1, 2.3, "5"), Tuple(1, 2.3, "4"))
+    XCTAssertNotEqual(Tuple(1, 2.9, "4"), Tuple(1, 2.3, "4"))
+    XCTAssertNotEqual(Tuple(0, 2.3, "4"), Tuple(1, 2.3, "4"))
+    // This effectively tests Hashable too; we know the conformance synthesizer
```

Comment:
I'll take your word on this one...

Reply:
By testing comparison we've verified that synthesized conformance is considering all the values the tuple is initialized with. Doing more would amount to testing the compiler, which is not what we need to do here.